perfkitbenchmarker/providers/gcp/gce_network.py

```diff
@@ -235,27 +243,72 @@ def __init__(self, network_spec):
     super(GceNetwork, self).__init__(network_spec)
     self.project = network_spec.project
     name = FLAGS.gce_network_name or 'pkb-network-%s' % FLAGS.run_uri
-    mode = 'auto' if FLAGS.gce_subnet_region is None else 'custom'
+
+    self.net_type = 'default'  # Default network 'auto' mode. Single VPC with subnets in all regions
+    self.subnet_addr = NETWORK_RANGE
+    self.net_name = network.BaseNetwork.FormatCidrString(NETWORK_RANGE)
+
+    if FLAGS.gce_subnet_region:  # Overrides auto network when set. Creates one VPC network in single region.
+      self.net_type = 'single'
+      self.subnet_addr = FLAGS.gce_subnet_addr
+      self.net_name = network.BaseNetwork.FormatCidrString(FLAGS.gce_subnet_addr)
+      name = FLAGS.gce_network_name or 'pkb-network-%s-%s-%s' % (self.net_type, self.net_name, FLAGS.run_uri)
+    if network_spec.cidr:  # Creates multiple VPC networks when set.
+      self.net_type = 'multi'
+      self.subnet_addr = network_spec.cidr
+      self.net_name = network.BaseNetwork.FormatCidrString(network_spec.cidr)
+      name = FLAGS.gce_network_name or 'pkb-network-%s-%s-%s' % (self.net_type, self.net_name, FLAGS.run_uri)
+
+    self.subnet_region = FLAGS.gce_subnet_region if network_spec.cidr is None else util.GetRegionFromZone(network_spec.zone)
+    mode = 'auto' if self.subnet_region is None else 'custom'
     self.network_resource = GceNetworkResource(name, mode, self.project)
-    if FLAGS.gce_subnet_region is None:
+    if self.subnet_region is None:
       self.subnet_resource = None
     else:
       self.subnet_resource = GceSubnetResource(name, name,
-                                               FLAGS.gce_subnet_region,
-                                               FLAGS.gce_subnet_addr,
+                                               self.subnet_region,
+                                               self.subnet_addr,
                                                self.project)
-    firewall_name = 'default-internal-%s' % FLAGS.run_uri
+
+    self.all_nets = []  # Holds the different networks we learn about.
+    self.external_nets_rules = {}  # Holds FW rules for any external subnets.
+    self.default_subnet = None  # Holds the default subnet for this run if used.
+
+    firewall_name = '%s-%s-%s-%s' % (self.net_type, 'internal', self.net_name, FLAGS.run_uri)
     self.default_firewall_rule = GceFirewallRule(
-        firewall_name, self.project, ALLOW_ALL, name, NETWORK_RANGE)
+        firewall_name, self.project, ALLOW_ALL, name, self.subnet_addr)
+    self.all_nets.append(self.net_name)
+    if self.net_type != 'multi':  # We're using the default network.
+      self.default_subnet = self.subnet_addr
+
+    # Add firewall rules for other networks if they exist.
```

Comment:
logic here is a bit hard to understand.
please break into helper method, add more comments and some tests if possible

Reply:
Fixed. 
I added the test file to tests/providers/gcp/ directory. 
There's a tests/gce_virtual_machine.py file that contained a couple firewall tests, so I copied them into this new file as well. 
I can merge all these tests back into the gce_virtual_machine.py tests if preferred... it wasn't clear to me why that file wasn't located in the providers/gcp directory.